aws_sigv4: merge repeated headers in canonical request

austinmoore- opened the pull request:
When multiple headers share the same name, AWS SigV4 expects them to be merged into a single header line in the canonical request, with values comma-delimited in the order they appeared.

This PR fixes the issue by combining repeated headers while creating the canonical request.

The fix has been verified with a few real S3 requests, along with the added libcurl test.

---

Detailed example:

```sh
curl -X PUT --verbose --silent \
    --aws-sigv4 "aws:amz:us-east-1:s3" --user "$AWS_ACCESS_KEY_ID:$AWS_SECRET_ACCESS_KEY" \
    -H "x-amz-meta-test: test2" \
    -H "x-amz-meta-test: test1" \
    --data-binary @testfile \
    "https://examplebucket.s3.amazonaws.com/testfile"
```

The canonical request that curl generates looks like:
```
PUT
/1B

host:examplebucket.s3.amazonaws.com
x-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date:20250317T054922Z
x-amz-meta-test:test2
x-amz-meta-test:test1

host;x-amz-content-sha256;x-amz-date;x-amz-meta-test
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

AWS will respond with a SignatureDoesNotMatch error because from the above request, it will generate the following as the canonical request:
```
PUT
/1B

host:examplebucket.s3.amazonaws.com
x-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date:20250317T054922Z
x-amz-meta-test:test2,test1

host;x-amz-content-sha256;x-amz-date;x-amz-meta-test
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

testclutch commented:
Analysis of PR #16743 at [bb50f7c7](https://github.com/curl/curl/commit/bb50f7c78ef119c21e727449ff372a266327890f):

[Test ../../tests/http/test_07_upload.py::TestUpload::test_07_42a_upload_disconnect[h2] failed](https://github.com/curl/curl/actions/runs/13892739199/job/38867344661#check-step-25), which has NOT been flaky recently, so **there could be a real issue in this PR**.

[Test 1978 failed](https://github.com/curl/curl/actions/runs/13892739184/job/38867352025#check-step-41), which has NOT been flaky recently, so **there could be a real issue in this PR**.

###### Generated by [Testclutch](https://testclutch.curl.se/)

vszakats commented:
> Analysis of PR #16743 at [bb50f7c7](https://github.com/curl/curl/commit/bb50f7c78ef119c21e727449ff372a266327890f):
> 
> [Test ../../tests/http/test_07_upload.py::TestUpload::test_07_42a_upload_disconnect[h2] failed](https://github.com/curl/curl/actions/runs/13892739199/job/38867344661#check-step-25), which has NOT been flaky recently, so **there could be a real issue in this PR**.
> 
> [Test 1978 failed](https://github.com/curl/curl/actions/runs/13892739184/job/38867352025#check-step-41), which has NOT been flaky recently, so **there could be a real issue in this PR**.
> 
> ###### Generated by [Testclutch](https://testclutch.curl.se/)

This seems legit:
```
test 1978...[HTTP AWS_SIGV4 canonical request duplicate header test]
 143 functions found, but only fail 25 (17.48%)
** MEMORY FAILURE
Leak detected: memory still allocated: 39 bytes
At 55a2208a0328, there's 16 bytes.
 allocated by /home/runner/work/curl/curl/lib/slist.c:68
At 55a2208a02f8, there's 23 bytes.
 allocated by /home/runner/work/curl/curl/lib/slist.c:94
LIMIT /home/runner/work/curl/curl/lib/slist.c:94 strdup reached memlimit
 1978: torture FAILED: function number 11 in test.
 invoke with "-t11" to repeat this single case.
```
https://github.com/curl/curl/actions/runs/13892739184/job/38867352025?pr=16743#step:41:4062

austinmoore- commented:
Ack, 1978 is the test that I added for this. I won't be able to look at it today, but I'll take a look to see what I missed freeing soon.

austinmoore- commented:
I was incorrectly creating the header list in the test, which has been addressed in the latest commit.

I'm still unsure about the [test_07_42a_upload_disconnect[h2]](https://github.com/curl/curl/actions/runs/13892739199/job/38867344661#check-step-25) failure. From what I can tell, there is no aws_sigv4 functionality in that test. I'm not super familiar with curl's source, but the changes in this PR are isolated to aws_sigv4 functionality and AFAIK shouldn't have any affect on HTTP2 functionality.

austinmoore- commented:
The job runner for [Windows / mingw, CM clang-x86_64 gnutls (pull_request)](https://github.com/curl/curl/actions/runs/13938312745/job/39010396822?pr=16743) exceeded the maximum execution time.

---

Also, I'm squashing the fix commit.

bagder commented on lib/http_aws_sigv4.c:
```diff
@@ -154,6 +154,66 @@ static int compare_header_names(const char *a, const char *b)
+      size_t val_next_len;
+      size_t new_len;
+
+      colon_next = strchr(next->data, ':');
```
maybe add a test that provides a header without a colon? Seems likely to be able to happen and an assert is not going to catch that in a release build.

austinmoore- replied:
Added to test

bagder commented on lib/http_aws_sigv4.c:
```diff
@@ -154,6 +154,66 @@ static int compare_header_names(const char *a, const char *b)
+      DEBUGASSERT(colon_next);
+
+      curr_len = strlen(curr->data);
+      val_next = colon_next + 1;
```
If `colon_next` is NULL here, this goes wrong.

austinmoore- replied:
I can't find a way that we'd get a header without a colon in here. By this point, all the headers should have a `:`.

I figured `DEBUGASSERT` would be enough to ensure this invariant during dev builds. However, if you'd prefer that it be changed to something like 
```
if(!colon_next)
    return /* some CURLcode */
```

I have no issues with that. Just lmk if you'd prefer that and which error code is best here (maybe `CURLE_BAD_FUNCTION_ARGUMENT`?)

bagder commented on lib/http_aws_sigv4.c:
```diff
@@ -154,6 +154,66 @@ static int compare_header_names(const char *a, const char *b)
+      /* add 1 for the comma we'll insert */
+      new_len = curr_len + 1 + val_next_len;
+
+      merged = malloc(new_len + 1); /* add 1 for null terminator */
```
I would prefer if this was written to use a `struct dynbuf` instead to build the new string in, to avoid the malloc and direct memcpy calls.

austinmoore- replied:
Updated to use dynbuf

bagder reviewed (commented):
Lovely take and bonus points for the awesome test case! Just some minor remarks from me.

austinmoore- commented:
Updated to use `dynbuf` in place of malloc/strcpy. I've also added a more headers to the test case.

Regarding the test case, let me know if its preferred they be split out. I went with the thinking of "test case for this fix" instead of "test case for failure situations" but I'm happy to split it if the latter is preferred.

bagder commented:
Thanks!
